extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py

```diff
@@ -0,0 +1,295 @@
+            CallbackEvent("on_run_complete", __name__, {"result": prompt.outputs})
+        )
+        return prompt.outputs
+
+    def get_output_text(
+        self,
+        prompt: Prompt,
+        aiconfig: "AIConfigRuntime",
+        output: Optional[Output] = None,
+    ) -> str:
+        if not output:
+            output = aiconfig.get_latest_output(prompt)
+
+        if not output:
+            return ""
+
+        if output.output_type == "execute_result":
+            output_data = output.data
+            if isinstance(output_data, str):
+                return output_data
+            else:
+                raise Exception("Not Implemented")
+
+        else:
+            return ""
+
+    def _construct_chat_history(
+        self, prompt: Prompt, aiconfig: "AIConfigRuntime", params: Dict
+    ) -> List:
+        """
+        Constructs the chat history for the model
+        """
+        messages = []
+        # Default to always use chat context
+        remember_chat_context = not hasattr(
+            prompt.metadata, "remember_chat_context"
+        ) or (
+            hasattr(prompt.metadata, "remember_chat_context")
+            and prompt.metadata.remember_chat_context != False
+        )
+        if remember_chat_context:
+            # handle chat history. check previous prompts for the same model. if same model, add prompt and its output to completion data if it has a completed output
+            for i, previous_prompt in enumerate(aiconfig.prompts):
+                # include prompts upto the current one
+                if previous_prompt.name == prompt.name:
+                    break
+
+                previous_prompt_is_same_model = aiconfig.get_model_name(
+                    previous_prompt
+                ) == aiconfig.get_model_name(prompt)
+                if previous_prompt_is_same_model:
+                    previous_prompt_template = resolve_prompt(
+                        previous_prompt, params, aiconfig
+                    )
+                    previous_prompt_output = aiconfig.get_latest_output(previous_prompt)
+                    previous_prompt_output_text = self.get_output_text(
+                        previous_prompt, aiconfig, previous_prompt_output
+                    )
+
+                    messages.append(
+                        {"role": "user", "parts": [{"text": previous_prompt_template}]}
+                    )
+                    messages.append(
+                        {
+                            "role": "model",
+                            "parts": [{"text": previous_prompt_output_text}],
+                        }
+                    )
+
+        return messages
+
+
+def refine_chat_completion_params(model_settings):
+    # completion parameters to be used for Palm's chat completion api
+    # messages handled seperately
+    supported_keys = {
+        "contents",
+        "generation_config",
+        "safety_settings",
+        "stream",
```

Comment:
'tools' can be added later on as well. For generation_config, do we want that to be sub-keyed under model_settings, e.g.
```
model.settings = { generation_config: {topK: ..., etc.} }
```

Or serialized top-level in the settings:
```
model.settings = {topK: ...}
```

The web app is doing the latter I think (cc @saqadri ).

Do we need to match camelCase of https://cloud.google.com/vertex-ai/docs/generative-ai/model-reference/gemini#gemini-pro or does the api here support this snake case?